src/main/java/net/imglib2/algorithm/hough/HoughTransforms.java

```diff
@@ -0,0 +1,343 @@
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ * #L%
+ */
+
+package net.imglib2.algorithm.hough;
+
+import java.util.List;
+
+import net.imglib2.Cursor;
+import net.imglib2.Point;
+import net.imglib2.RandomAccess;
+import net.imglib2.RandomAccessibleInterval;
+import net.imglib2.algorithm.localextrema.LocalExtrema;
+import net.imglib2.algorithm.localextrema.LocalExtrema.MaximumCheck;
+import net.imglib2.type.NativeType;
+import net.imglib2.type.numeric.RealType;
+import net.imglib2.util.Util;
+import net.imglib2.view.IntervalView;
+import net.imglib2.view.Views;
+
+/**
+ * This abstract class provides some basic functionality for use with arbitrary
+ * Hough-like transforms.
+ *
+ * @param <T>
+ *            the data type of the input image.
+ * @author Larry Lindsey
+ * @author Yili Zhao
+ * @author Gabe Selzer
+ */
+public class HoughTransforms< T extends RealType< T > & Comparable< T > >
+{
+
+	public static final int DEFAULT_THETA = 180;
+
+	/**
+	 * Calculates the geometric distance between [0, ... , 0] and position.
+	 * 
+	 * @param position
+	 *            - the position of the second point
+	 * @return {@code float} - the distance between the two points.
+	 */
+	final private static float computeLength( final long[] position )
+	{
+		float dist = 0;
+
+		for ( int d = 0; d < position.length; ++d )
+		{
+			final long pos = position[ d ];
+
+			dist += pos * pos;
+		}
+
+		return ( float ) Math.sqrt( dist );
+	}
+
+	/**
+	 * Calculates a default number of rho bins, which corresponds to a
+	 * resolution of one pixel.
+	 *
+	 * @param inputImage
+	 *            the {@link RandomAccessibleInterval} in question.
+	 * @return default number of rho bins.
+	 */
+	private static int defaultRho( final RandomAccessibleInterval< ? > inputImage )
+	{
+		final long[] dims = new long[ inputImage.numDimensions() ];
+		inputImage.dimensions( dims );
+		return ( int ) ( 2 * computeLength( dims ) );
+	}
+
+	/**
+	 * Returns the size of the vote space output image given an input
+	 * {@link RandomAccessibleInterval}.
+	 * 
+	 * @param input
+	 *            - the {@link RandomAccessibleInterval} over which the Hough
+	 *            Line Transform will be run
+	 * @return {@code long[]} - the dimensions of the vote space image
+	 */
+	public static < T extends RealType< T > > long[] getVotespaceSize( RandomAccessibleInterval< T > input )
```

Comment:
replace `final RAI inputImage` with `final Dimensions dimensions`, no need for generic parameter `T`.